Implement wNAF-based MSM derived from Gemini

Pratyush opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before hitting that submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: #450 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

Pratyush commented:
Hm this is giving an incorrect result. Any ideas @mmaker?

mmaker commented:
Hey! Thank you so much for bringing this into algebra. I'm unlikely to help review and fix this properly before the 20th December. Skimming though it, I didn't find any blatant error --   In Gemini this was still giving the same result  last time I tried?

Pratyush commented:
No worries! I think I've fixed the bug; I had updated how buckets are accessed when the digit was negative, and had done so incorrectly. It's fixed now, I believe.
